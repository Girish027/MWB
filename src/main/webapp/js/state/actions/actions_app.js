
import _ from 'lodash';
import getUrl, { pathKey } from 'utils/apiUrls';
import { fetchUtil } from 'utils/fetchUtils';
import goToRoute from 'utils/routeHelpers';
import Constants from 'constants/Constants';
import AmplitudeConstants from 'constants/AmplitudeConstants';
import { logAmplitudeEvent } from 'utils/amplitudeUtils';
import axios from 'axios';
import api from 'utils/api';
import { setDefaultsForAPI } from 'utils/api';
import * as types from './types';

// Note to developers:
// Both Warning and Good Messages are shown for the period of notificationInterval.
// Bad, Persistant - till user closes the notification. This should only be used sparingly.
export const callLogIngestAPI = (message, logLevel) => {
  const promise = new Promise((resolve, reject) => {
    const url = getUrl(pathKey.logIngest, { logLevel });
    return new Promise((resolve, reject) => {
      api.post({
        url,
        headers: {
          'Content-Type': 'text/plain',
        },
        data: message,
      });
    });
  });
};

export const displayBadRequestMessage = (message) => (dispatch) => {
  dispatch({
    type: types.BAD_REQUEST,
    value: message,
  });

  callLogIngestAPI(message, Constants.LOG_LEVEL_ERROR);
};

export const displayWarningRequestMessage = (message, notificationInterval = Constants.NOTIFICATION.interval) => (dispatch) => {
  dispatch({
    type: types.WARNING_REQUEST,
    value: message,
  });

  callLogIngestAPI(message, Constants.LOG_LEVEL_WARNING);

  setTimeout(() => {
    dispatch(stopShowingServerMessage());
  }, notificationInterval);
};

export const displayGoodRequestMessage = (message, notificationInterval = Constants.NOTIFICATION.interval) => (dispatch) => {
  dispatch({
    type: types.GOOD_REQUEST,
    value: message,
  });

  setTimeout(() => {
    dispatch(stopShowingServerMessage());
  }, notificationInterval);
};

export const displayPersistantMessage = message => ({
  type: types.PERSISTENT_REQUEST,
  value: message,
});

export const stopShowingServerMessage = () => ({
  type: types.APP_STOP_SHOWING_SERVER_MESSAGE,
});

export const updateLoginSuccess = csrfToken => ({
  type: types.LOGIN_SUCCESS,
  value: csrfToken,
});

export const updateLoginInfo = (userId, userName, userDetails) => ({
  type: types.LOGIN_INFO,
  value: {
    userId,
    userName,
    userDetails,
  },
});

export const logoutFromOktaSuccess = () => ({
  type: types.LOGOUT_OKTA,
});

export const logoutFromMWBSuccess = () => ({
  type: types.LOGOUT_MWB,
});

export const contextMenuChange = contextMenuState => ({
  type: types.CONTEXT_MENU_CHANGE,
  contextMenuState,
});

export const modalDialogChange = modalDialogState => ({
  type: types.MODAL_DIALOG_CHANGE,
  modalDialogState,
});

const actionFileRemoveRequest = ({ fileId }) => ({
  type: types.FILE_REMOVE_REQUEST,
  fileId,
});

const actionFileRemoveError = ({ fileId, error }) => ({
  type: types.FILE_REMOVE_ERROR,
  fileId,
  error,
});

const actionFileRemoveSuccess = ({ fileId }) => ({
  type: types.FILE_REMOVE_SUCCESS,
  fileId,
});

export const updateUserGroups = data => ({
  type: types.UPDATE_USER_GROUPS,
  data,
});

const onFetchFailureRemoveFile = (error, params) => (dispatch) => {
  const { fileId } = params;
  dispatch(actionFileRemoveError({ fileId, error }));
};

const onFetchSuccessRemoveFile = (data, params) => (dispatch) => {
  const { fileId } = params;
  dispatch(actionFileRemoveSuccess({ fileId }));
};

export const removeFile = ({ fileId }) => (dispatch, getState) => {
  dispatch(actionFileRemoveRequest({ fileId }));

  const state = getState();
  const fetchUrl = getUrl(pathKey.files, { fileId, clientId: state.header.client.id });

  fetchUtil({
    fetchUrl,
    fetchMethod: 'delete',
    onFetchSuccess: onFetchSuccessRemoveFile,
    onFetchFailure: onFetchFailureRemoveFile,
    dispatch,
    getState,
    params: {
      fileId,
    },
  });
};

const onFetchSuccessGetUserGroups = data => (dispatch) => {
  dispatch(updateUserGroups(data));
};

export const getUserGroups = ({ csrfToken }) => (dispatch, getState) => {
  const fetchUrl = getUrl(pathKey.userGroups, { csrfToken });

  fetchUtil({
    fetchUrl,
    fetchMethod: 'get',
    onFetchSuccess: onFetchSuccessGetUserGroups,
    dispatch,
    getState,
  });
};

export const onFetchSuccessGetCSRFToken = data => (dispatch, getState) => {
  const csrfToken = data.value;
  setDefaultsForAPI(csrfToken);
  dispatch(updateLoginSuccess(csrfToken));
  dispatch(getUserGroups({ csrfToken }));
};

export const getCSRFToken = userId => (dispatch, getState) => {
  const fetchUrl = getUrl(pathKey.csrfToken);

  fetchUtil({
    fetchUrl,
    fetchMethod: 'get',
    onFetchSuccess: onFetchSuccessGetCSRFToken,
    dispatch,
    getState,
    params: {
      userId,
    },
  });
};

const logoutFromMWB = () => (dispatch, getState) => {
  dispatch(logoutFromOktaSuccess());
  dispatch(logoutFromMWBSuccess());
};

export const logoutFromOkta = () => (dispatch, getState) => {
  const state = getState();
  const { oAuthLogoutURL } = state.app;
  if (oAuthLogoutURL) {
    fetch(oAuthLogoutURL, {
      method: 'DELETE',
      credentials: 'include',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
    })
      .then((response) => {
        window.location.href = Constants.LOGOUT;
        switch (response.status) {
        case 204:
          logoutFromMWB();
          logAmplitudeEvent(AmplitudeConstants.LOGOUT_EVENT, state);
          break;
        case 404:
        default:
          // TODO send log for these scenarios
          break;
        }
      })
      .catch((error) => {
        // TODO send log for these scenarios
      });
  }
};

export const changeRoute = (routeName, data, history) => (dispatch, getState) => {
  const state = getState();
  const routeData = Object.assign({}, {
    client: !_.isNil(state.header.client) ? state.header.client : {
      id: 0,
      name: 'Loading...',
    },
    projectId: state.selectedProjectId,
  }, data);
  goToRoute(routeName, routeData, history);
};
