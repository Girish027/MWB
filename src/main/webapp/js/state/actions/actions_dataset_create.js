import fetch from 'isomorphic-fetch';
import errorMessageUtil from 'utils/ErrorMessageUtil';
import { addDatasetToProject, fetchDatasetTransform } from 'state/actions/actions_datasets';
import * as apiUtils from 'utils/apiUtils';
import getUrl, { pathKey } from 'utils/apiUrls';
import { getLanguage } from 'state/constants/getLanguage';
import { logAmplitudeEvent } from 'utils/amplitudeUtils';
import AmplitudeConstants from 'constants/AmplitudeConstants';
import Constants from 'constants/Constants';
import api from 'utils/api';
import * as projectsActions from 'state/actions/actions_projects';
import { escapeRegExp } from 'utils/StringUtils';
import * as appActions from './actions_app';
import * as types from './types';
import { colors } from '../../styles';

const languageConstants = getLanguage();
const { DISPLAY_MESSAGES } = languageConstants;

export const reset = () => ({
  type: types.CREATE_DATASET_DIALOG_RESET,
});

export const fileUploadSuccess = ({ token, columns, previewData }) => ({
  type: types.CREATE_DATASET_DIALOG_FILE_UPLOAD_SUCCESS,
  token,
  columns,
  previewData,
});

export const changeStep = ({ step }) => ({
  type: types.CREATE_DATASET_DIALOG_CHANGE_STEP,
  step,
});

export const columnsBind = ({
  bindingArray, isBindingValid, columnsBinding, isPreSelected,
}) => ({
  type: types.CREATE_DATASET_DIALOG_COLUMNS_BIND,
  bindingArray,
  isBindingValid,
  columnsBinding,
  isPreSelected,
});

export const changeAutoTag = ({ autoTag }) => ({
  type: types.CREATE_DATASET_DIALOG_AUTO_TAG_CHANGE,
  autoTag,
});

export const changesStartTransform = ({ startTransform }) => ({
  type: types.CREATE_DATASET_DIALOG_START_TRANSFORM_CHANGE,
  startTransform,
});

export const changeFirstRowSkip = ({ skipFirstRow }) => ({
  type: types.CREATE_DATASET_DIALOG_FISRT_ROW_SKIP_CHANGE,
  skipFirstRow,
});

const columnsBindRequest = ({ token, bindingArray, skipFirstRow }) => ({
  type: types.CREATE_DATASET_DIALOG_COLUMNS_BIND_REQUEST,
  token,
  bindingArray,
  skipFirstRow,
});

const columnsBindError = ({
  token, bindingArray, skipFirstRow, error,
}) => ({
  type: types.CREATE_DATASET_DIALOG_COLUMNS_BIND_ERROR,
  token,
  bindingArray,
  skipFirstRow,
  error,
});

const columnsBindSuccess = ({
  token, bindingArray, skipFirstRow, fileId,
}) => ({
  type: types.CREATE_DATASET_DIALOG_COLUMNS_BIND_SUCCESS,
  token,
  bindingArray,
  skipFirstRow,
  fileId,
});

export const requestColumnsBind = ({ token, bindingArray, skipFirstRow }) => (dispatch, getState) => new Promise((resolve, reject) => {
  dispatch(columnsBindRequest({ token, bindingArray, skipFirstRow }));

  const state = getState();
  const { csrfToken } = state.app;
  const clientId = state.header.client.id;

  const fetchUrl = getUrl(pathKey.importDatasetColumnsBind, { token, skipFirstRow, clientId });

  fetch(fetchUrl, {
    credentials: 'same-origin',
    method: 'POST',
    headers: {
      'X-CSRF-TOKEN': csrfToken,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(bindingArray),
  })
    .then(errorMessageUtil.handleErrors)
    .then(response => response.json())
    .then((response) => {
      /* fileId, name, systemName */
      const bindData = Object.assign({}, {
        token,
        bindingArray,
        skipFirstRow,
        fileId: response.fileId,
      });
      dispatch(columnsBindSuccess(bindData));
      resolve(bindData);
    })
    .catch((error) => {
      dispatch(columnsBindError({
        token, bindingArray, skipFirstRow, error,
      }));
      dispatch(appActions.modalDialogChange(null));
      errorMessageUtil.dispatchError(error, dispatch, appActions.displayBadRequestMessage);
      reject(error);
      dispatch(reset());
    });
});

const createDatasetRequest = () => ({
  type: types.CREATE_DATASET_REQUEST,
});

const createDatasetError = () => ({
  type: types.CREATE_DATASET_ERROR,
});

const createDatasetSuccess = () => ({
  type: types.CREATE_DATASET_SUCCESS,
});

export const createDataset = ({
  projectId, clientId, dataType, locale, name, uri, description, autoTagDataset, mapToProjectId,
}) => (dispatch, getState) => new Promise((resolve, reject) => {
  dispatch(createDatasetRequest());

  const state = getState();
  const { userDetails = Constants.OBJECT_DEFAULT_VALUE, csrfToken } = state.app;
  const { userType = Constants.DEFAULT_VALUE } = userDetails;
  const fetchUrl = getUrl(pathKey.datasets, { clientId });
  let source = Constants.DATASET_EXTERNAL_FLAG;

  logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_EVENT, state, {
    projectId, clientDBId: clientId,
  });

  if (userType == Constants.USER_TYPE_INTERNAL) {
    source = Constants.DATASET_INTERNAL_FLAG;
  }

  fetch(fetchUrl, {
    method: 'POST',
    credentials: 'same-origin',
    headers: {
      'X-CSRF-TOKEN': csrfToken,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      dataset: {
        projectId, clientId, dataType, locale, name, source, uri, description,
      },
      projectId,
      autoTagDataset,
    }),
  })
    .then(errorMessageUtil.handleErrors)
    .then(response => response.json())
    .then((response) => {
      const newDataset = apiUtils.normalizeIds(response);
      const dataset = Object.assign({ status: 'NULL', type: newDataset.dataType, description: newDataset.description || '' }, newDataset);
      dispatch(createDatasetSuccess());
      if (!mapToProjectId) {
        resolve(dataset);
      } else {
        dispatch(addDatasetToProject({
          datasetId: dataset.id,
          projectId,
          dataset,
        }))
          .then(() => {
            logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_COMPLETED_EVENT, state, {
              projectId, datasetId: dataset.id, clientDBId: clientId,
            });
            dispatch(projectsActions.refreshProjectsByClient(clientId));
            resolve(dataset);
          })
          .catch((error) => {
            logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
              projectId, clientDBId: clientId,
            });
            dispatch(appActions.modalDialogChange(null));
            reject(error);
          });
      }
    })
    .catch((error) => {
      dispatch(appActions.modalDialogChange(null));
      switch (error.status) {
      case 400:
        logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
          projectId, clientDBId: clientId, source, reason: DISPLAY_MESSAGES.createFailed,
        });
        dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.createFailed));
        break;
      case 403:
        logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
          projectId, clientDBId: clientId, source, reason: DISPLAY_MESSAGES.datasetOperationNotAllowed,
        });
        dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.datasetOperationNotAllowed));
        break;
      case 409:
        logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
          projectId, clientDBId: clientId, source, reason: DISPLAY_MESSAGES.datasetAlreadyExists(name),
        });
        dispatch(appActions.displayWarningRequestMessage(DISPLAY_MESSAGES.datasetAlreadyExists(name)));
        break;
      default:
        logAmplitudeEvent(AmplitudeConstants.CREATE_DATASET_FAILED_EVENT, state, {
          projectId, clientDBId: clientId, source, reason: DISPLAY_MESSAGES.datasetCreateError,
        });
        dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.datasetCreateError));
        errorMessageUtil.dispatchError(error, dispatch, appActions.displayBadRequestMessage);
      }
      dispatch(createDatasetError());
      reject(error);
    });
});

export const onDatasetImportSuccess = (response = {}, importData) => (dispatch, getState) => {
  let { token, columns, previewData } = response;
  if (previewData && previewData.length > 10) {
    previewData = previewData.slice(0, 10);
  }
  dispatch(fileUploadSuccess({ token, columns, previewData }));
  dispatch(changeStep({ step: 'mapping' }));
  dispatch(columnBinding(importData));
};

export const onDatasetImportFailure = (error = {}) => (dispatch, getState) => {
  const { code } = error;
  if (code == 400) {
    const { message } = error;
    dispatch(appActions.displayBadRequestMessage(message));
  } else if (code == 401) {
    dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.sessionExpired));
  } else {
    dispatch(appActions.displayBadRequestMessage(DISPLAY_MESSAGES.fileParseFail));
  }
  dispatch(reset());
  dispatch(appActions.modalDialogChange(null));
};

export const importFile = importData => (dispatch, getState) => {
  const { dropZone } = importData;
  const url = getUrl(pathKey.importDataset);
  const formData = new FormData();
  if (dropZone) {
    formData.append('datatype', null);
    formData.append(Constants.DATASET_FILE, dropZone, dropZone.name);
  }

  return new Promise((resolve, reject) => {
    api.post({
      dispatch,
      getState,
      url,
      data: formData,
      onApiSuccess: response => onDatasetImportSuccess(response, importData),
      onApiError: onDatasetImportFailure,
    });
  });
};

export const columnBinding = importData => (dispatch, getState) => {
  const state = getState();
  const { createDatasetDialog } = state;
  const {
    columns, previewData, token,
    isMappingRequestLoading,
  } = createDatasetDialog;
  const { isSomeColumnsPreSelected, columnsBinding } = tryPreSelectColumns(columns,
    previewData, state.createDatasetDialog.columnsBinding);

  let requiredColumns = [];
  let columnsByName = {};

  columns.forEach(column => {
    columnsByName[column.name] = column;
    if (column.required || column.name === 'rutag') {
      requiredColumns.push(column.name);
    }
  });

  if (isSomeColumnsPreSelected) {
    let isBindingValid = true;
    requiredColumns.forEach(column => {
      if (typeof columnsBinding[column] === 'undefined') {
        isBindingValid = false;
      }
    });

    const bindingArray = [];
    for (const columnName in columnsBinding) {
      if (!columnsBinding.hasOwnProperty(columnName)) {
        continue;
      }
      const column = columnsByName[columnName];
      bindingArray.push({
        id: `${column.id}`,
        columnName: column.name,
        columnIndex: `${columnsBinding[columnName]}`,
        displayName: column.displayName,
      });
    }

    isBindingValid = isBindingValid && bindingArray.length > 0;

    dispatch(columnsBind({
      bindingArray, isBindingValid, columnsBinding, isPreSelected: true,
    }));

    if (isBindingValid && bindingArray.length > 0 && token && !isMappingRequestLoading) {
      dispatch(confirmDatasetUpload(importData));
    } else {
      dispatch(showDatasetMapperDialog(importData));
    }
  } else {
    dispatch(showDatasetMapperDialog(importData));
  }
};

const showDatasetMapperDialog = importData => (dispatch, getState) => {
  const state = getState();
  const { createDatasetDialog } = state;
  dispatch(appActions.modalDialogChange({
    dispatch,
    createDatasetDialog,
    type: Constants.DIALOGS.DATASET_COLUMN_MAPPING_DIALOG,
    onOk: () => dispatch(confirmDatasetUpload(importData)),
  }));
};

const confirmDatasetUpload = importData => (dispatch, getState) => {
  const state = getState();
  const { createDatasetDialog, app } = state;
  const { userId, csrfToken } = app;
  const { token, bindingArray, skipFirstRow } = createDatasetDialog;
  const {
    name, description, projectId,
    clientId, locale, dataType,
  } = importData;
  dispatch(appActions.modalDialogChange({
    dispatch,
    type: Constants.DIALOGS.PROGRESS_DIALOG,
    message: Constants.UPLOAD_DATASET_PROGRESS,
    header: Constants.UPLOADING_DATASET,
    showHeader: true,
    showFooter: true,
    cancelVisible: true,
    closeIconVisible: true,
    cancelChildren: Constants.CLOSE,
    styleOverride: {
      childContainer: {
        marginTop: '30px',
        marginBottom: '10px',
      },
      content: {
        top: '160px',
        maxWidth: '444px',
        maxHeight: '333px',
        left: 'calc((100vw - 500px) / 2)',
      },
      cancel: {
        color: `${colors.white}`,
        backgroundColor: `${colors.cobalt}`,
        paddingLeft: '25px',
        paddingRight: '25px',
        ':hover': {
          color: `${colors.white}`,
          backgroundColor: `${colors.prussianBlue}`,
        },
      },
    },
  }));
  dispatch(requestColumnsBind({
    token, bindingArray, skipFirstRow,
  })).then((bindData) => {
    const { fileId } = bindData;
    dispatch(createDataset({
      projectId,
      clientId,
      dataType,
      locale,
      name,
      uri: `https://tagging.247-inc.com:8443/nltools/private/v1/files/${fileId}`,
      description,
      autoTagDataset: true,
      mapToProjectId: projectId,
    })).then((dataset) => {
      setTimeout(() => {
        dispatch(fetchDatasetTransform(userId, dataset.id, clientId, projectId, csrfToken, false));
      }, 2000);
      dispatch(appActions.displayGoodRequestMessage(DISPLAY_MESSAGES.datasetCreated(dataset.name)));
      dispatch(reset());
    });
  });
};

const tryPreSelectColumns = (columns, data, columnsBindingInput) => {
  const firstRow = data.length ? data[0] : [];
  const columnsBinding = Object.assign({}, columnsBindingInput);
  if (!firstRow.length || !columns.length || (Object.keys(columnsBinding)).length != 0) {
    return { isSomeColumnsPreSelected: false, columnsBinding };
  }

  columns.forEach((c) => {
    let foundIndexes = [];
    let usedIndexes = {};

    // pass 1 - exact match
    let tests = [
      new RegExp(`^${escapeRegExp(c.name)}$`, 'i'),
      new RegExp(`^${escapeRegExp(c.displayName)}$`, 'i'),
    ];

    firstRow.forEach((newData1, index) => {
      for (let i = 0; i < tests.length; i++) {
        if (tests[i].test(newData1.trim())) {
          foundIndexes.push(index);
          break;
        }
      }
    });

    if (!foundIndexes.length) {
      // pass 2 - partial match
      tests = [
        new RegExp(escapeRegExp(c.name), 'i'),
        new RegExp(escapeRegExp(c.displayName), 'i'),
      ];

      firstRow.forEach((newData2, index) => {
        for (let i = 0; i < tests.length; i++) {
          if (tests[i].test(newData2)) {
            foundIndexes.push(index);
            break;
          }
        }
      });
    }

    if (foundIndexes.length === 1 && !usedIndexes[foundIndexes[0]]) {
      columnsBinding[c.name] = foundIndexes[0];
      usedIndexes[foundIndexes[0]] = true;
    }
  });
  return { isSomeColumnsPreSelected: (Object.keys(columnsBinding)).length != 0, columnsBinding };
};

export const fileUpload = (file, handleSuccess, handleError) => (dispatch, getState) => {
  const fetchUrl = getUrl(pathKey.importDataset);
  const formData = new FormData();
  formData.append('datatype', null);
  formData.append('file', file);
  let req = new XMLHttpRequest();
  req.open('post', fetchUrl, true);
  req.onload = (event) => {
    let { response } = req;
    if (req.status == 200) {
      handleSuccess(response);
    } else {
      handleError(response);
    }
  };
  req.send(formData);
};
