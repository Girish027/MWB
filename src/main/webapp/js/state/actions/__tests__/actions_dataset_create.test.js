jest.mock('utils/api');

import api from 'utils/api';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as datasetCreateActions from 'state/actions/actions_dataset_create';

const middlewares = [thunk];
const reduxMockStore = configureMockStore(middlewares);


describe('actions_dataset_create', () => {
  let store;
  let file = new Blob([
    JSON.stringify({ name: 'dropzone' }),
  ], { type: 'application/json' });
  let importData = {
    dropZone: file[0],
    name: 'dataset',
    description: 'description',
    projectId: '123',
    clientId: '14',
    locale: 'en-US',
    dataType: 'Text',
  };
  let props = {
    handleSuccess: jest.fn(),
    handleError: jest.fn(),
  };
  let data = {
    token: 'token',
    columns: [
      {
        id: '1',
        name: 'transcription',
        required: true,
        displayName: 'Transcription',
      },
      {
        id: '3',
        name: 'filename',
        required: false,
        displayName: 'Filename',
      },
      {
        id: '5',
        name: 'inheritedIntent',
        required: false,
        displayName: 'Granular Intent',
      },
      {
        id: '7',
        name: 'rutag',
        required: false,
        displayName: 'Rollup Intent',
      },
      {
        id: '9',
        name: 'filesize',
        required: false,
        displayName: 'Filesize',
      },
      {
        id: '11',
        name: 'date',
        required: false,
        displayName: 'Date',
      },
      {
        id: '13',
        name: 'dataset',
        required: false,
        displayName: 'Dataset',
      },
    ],
    previewData: [
      [
        'Transcription',
        'Rollup Intent',
      ],
      [
        'What are store hours',
        'Store_Hours',
      ],
      [
        'store hours',
        'Store_Hours',
      ],
      [
        'store opening time',
        'Store_Hours',
      ],
      [
        'store closing time',
        'Store_Hours',
      ],
      [
        'store times',
        'Store_Hours',
      ],
      [
        'When does this shop open',
        'Store_Hours',
      ],
      [
        'store timings',
        'Store_Hours',
      ],
      [
        'What are store timings',
        'Store_Hours',
      ],
      [
        'Tell me the store hours',
        'Store_Hours',
      ],
      [
        'I would like to change password',
        'Password_Change',
      ],
    ],
  };

  beforeAll(() => {
    store = reduxMockStore({
      app: {
        userId: 'testUser',
      },
      header: {
        client: {
          id: '14',
        },
      },
      projectListSidebar: {
        projectById: {
          123: {
            name: 'OldName',
          },
        },
      },
      createDatasetDialog: {
        step: 'upload',
        columns: data.columns,
        previewData: data.previewData,
        bindingArray: [],
        columnsBinding: {},
        isPreSelected: false,
        isBindingValid: false,
        skipFirstRow: true,
        isMappingRequestLoading: false,
        isCommitRequestLoading: false,
        token: '8182f040-8c4c-40c5-b864-3ffb2d71d330',
        fileId: null,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
    store.clearActions();
  });

  describe('onDatasetImportSuccess', () => {
    test('should dispatch onDatasetImportSuccess action with data', () => {
      store.dispatch(datasetCreateActions.onDatasetImportSuccess(data, {}));
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('onDatasetImportFailure', () => {
    test('should dispatch onDatasetImportFailure action', () => {
      store.dispatch(datasetCreateActions.onDatasetImportFailure({}));
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  describe('importFile', () => {
    test('should dispatch importFile action', () => {
      store.dispatch(datasetCreateActions.importFile(importData));
      expect(store.getActions()).toMatchSnapshot();
    });

    test('should make a GET call to get the config properties', () => {
      store.dispatch(datasetCreateActions.importFile(importData));
      expect(api.post).toHaveBeenCalledTimes(1);
      expect(api.post.mock.calls[0][0].url).toMatchSnapshot();
    });

    test('should dispatch fileUpload action', () => {
      store.dispatch(datasetCreateActions.fileUpload(importData.dropZone, props.handleSuccess, props.handleError));
      expect(store.getActions()).toMatchSnapshot();
    });

    test('should dispatch columnBinding action', () => {
      store.dispatch(datasetCreateActions.columnBinding(importData));
      expect(store.getActions()).toMatchSnapshot();
    });

    test('should dispatch createDataset action', () => {
      const {
        projectId, clientId, dataType,
        locale, name, description,
      } = importData;
      const uri = 'https://tagging.247-inc.com:8443/nltools/private/v1/files';
      store.dispatch(datasetCreateActions.createDataset({
        projectId,
        clientId,
        dataType,
        locale,
        name,
        uri,
        description,
        autoTagDataset: true,
        mapToProjectId: true,
      }));
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
