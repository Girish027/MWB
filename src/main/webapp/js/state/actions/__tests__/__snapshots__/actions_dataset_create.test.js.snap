// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions_dataset_create importFile should dispatch columnBinding action 1`] = `
Array [
  Object {
    "bindingArray": Array [
      Object {
        "columnIndex": "0",
        "columnName": "transcription",
        "displayName": "Transcription",
        "id": "1",
      },
      Object {
        "columnIndex": "1",
        "columnName": "rutag",
        "displayName": "Rollup Intent",
        "id": "7",
      },
    ],
    "columnsBinding": Object {
      "rutag": 1,
      "transcription": 0,
    },
    "isBindingValid": true,
    "isPreSelected": true,
    "type": "CREATE_DATASET_DIALOG_COLUMNS_BIND",
  },
  Object {
    "modalDialogState": Object {
      "cancelChildren": "CLOSE",
      "cancelVisible": true,
      "closeIconVisible": true,
      "dispatch": [Function],
      "header": "Uploading your file…",
      "message": "You may close this window while we upload your file. Uploading will not cancel. We will prompt you once your file has successfully uploaded.",
      "showFooter": true,
      "showHeader": true,
      "styleOverride": Object {
        "cancel": Object {
          ":hover": Object {
            "backgroundColor": "#003467",
            "color": "#fff",
          },
          "backgroundColor": "#004c97",
          "color": "#fff",
          "paddingLeft": "25px",
          "paddingRight": "25px",
        },
        "childContainer": Object {
          "marginBottom": "10px",
          "marginTop": "30px",
        },
        "content": Object {
          "left": "calc((100vw - 500px) / 2)",
          "maxHeight": "333px",
          "maxWidth": "444px",
          "top": "160px",
        },
      },
      "type": "ProgressDialog",
    },
    "type": "MODAL_DIALOG_CHANGE",
  },
  Object {
    "bindingArray": Array [],
    "skipFirstRow": true,
    "token": "8182f040-8c4c-40c5-b864-3ffb2d71d330",
    "type": "CREATE_DATASET_DIALOG_COLUMNS_BIND_REQUEST",
  },
]
`;

exports[`actions_dataset_create importFile should dispatch createDataset action 1`] = `
Array [
  Object {
    "type": "CREATE_DATASET_REQUEST",
  },
]
`;

exports[`actions_dataset_create importFile should dispatch fileUpload action 1`] = `Array []`;

exports[`actions_dataset_create importFile should dispatch importFile action 1`] = `Array []`;

exports[`actions_dataset_create importFile should make a GET call to get the config properties 1`] = `"/nltools/private/v1/files/import"`;

exports[`actions_dataset_create onDatasetImportFailure should dispatch onDatasetImportFailure action 1`] = `
Array [
  Object {
    "type": "BAD_REQUEST",
    "value": "Error in parsing the data given in the uploaded file, please check documentation for expected format.",
  },
  Object {
    "type": "CREATE_DATASET_DIALOG_RESET",
  },
  Object {
    "modalDialogState": null,
    "type": "MODAL_DIALOG_CHANGE",
  },
]
`;

exports[`actions_dataset_create onDatasetImportSuccess should dispatch onDatasetImportSuccess action with data 1`] = `
Array [
  Object {
    "columns": Array [
      Object {
        "displayName": "Transcription",
        "id": "1",
        "name": "transcription",
        "required": true,
      },
      Object {
        "displayName": "Filename",
        "id": "3",
        "name": "filename",
        "required": false,
      },
      Object {
        "displayName": "Granular Intent",
        "id": "5",
        "name": "inheritedIntent",
        "required": false,
      },
      Object {
        "displayName": "Rollup Intent",
        "id": "7",
        "name": "rutag",
        "required": false,
      },
      Object {
        "displayName": "Filesize",
        "id": "9",
        "name": "filesize",
        "required": false,
      },
      Object {
        "displayName": "Date",
        "id": "11",
        "name": "date",
        "required": false,
      },
      Object {
        "displayName": "Dataset",
        "id": "13",
        "name": "dataset",
        "required": false,
      },
    ],
    "previewData": Array [
      Array [
        "Transcription",
        "Rollup Intent",
      ],
      Array [
        "What are store hours",
        "Store_Hours",
      ],
      Array [
        "store hours",
        "Store_Hours",
      ],
      Array [
        "store opening time",
        "Store_Hours",
      ],
      Array [
        "store closing time",
        "Store_Hours",
      ],
      Array [
        "store times",
        "Store_Hours",
      ],
      Array [
        "When does this shop open",
        "Store_Hours",
      ],
      Array [
        "store timings",
        "Store_Hours",
      ],
      Array [
        "What are store timings",
        "Store_Hours",
      ],
      Array [
        "Tell me the store hours",
        "Store_Hours",
      ],
    ],
    "token": "token",
    "type": "CREATE_DATASET_DIALOG_FILE_UPLOAD_SUCCESS",
  },
  Object {
    "step": "mapping",
    "type": "CREATE_DATASET_DIALOG_CHANGE_STEP",
  },
  Object {
    "bindingArray": Array [
      Object {
        "columnIndex": "0",
        "columnName": "transcription",
        "displayName": "Transcription",
        "id": "1",
      },
      Object {
        "columnIndex": "1",
        "columnName": "rutag",
        "displayName": "Rollup Intent",
        "id": "7",
      },
    ],
    "columnsBinding": Object {
      "rutag": 1,
      "transcription": 0,
    },
    "isBindingValid": true,
    "isPreSelected": true,
    "type": "CREATE_DATASET_DIALOG_COLUMNS_BIND",
  },
  Object {
    "modalDialogState": Object {
      "cancelChildren": "CLOSE",
      "cancelVisible": true,
      "closeIconVisible": true,
      "dispatch": [Function],
      "header": "Uploading your file…",
      "message": "You may close this window while we upload your file. Uploading will not cancel. We will prompt you once your file has successfully uploaded.",
      "showFooter": true,
      "showHeader": true,
      "styleOverride": Object {
        "cancel": Object {
          ":hover": Object {
            "backgroundColor": "#003467",
            "color": "#fff",
          },
          "backgroundColor": "#004c97",
          "color": "#fff",
          "paddingLeft": "25px",
          "paddingRight": "25px",
        },
        "childContainer": Object {
          "marginBottom": "10px",
          "marginTop": "30px",
        },
        "content": Object {
          "left": "calc((100vw - 500px) / 2)",
          "maxHeight": "333px",
          "maxWidth": "444px",
          "top": "160px",
        },
      },
      "type": "ProgressDialog",
    },
    "type": "MODAL_DIALOG_CHANGE",
  },
  Object {
    "bindingArray": Array [],
    "skipFirstRow": true,
    "token": "8182f040-8c4c-40c5-b864-3ffb2d71d330",
    "type": "CREATE_DATASET_DIALOG_COLUMNS_BIND_REQUEST",
  },
]
`;
